pkg/client/image.go

```diff
@@ -173,19 +175,48 @@ func (c *DefaultClient) ImageDelete(ctx context.Context, imageName string, opts
 	} else if err != nil {
 		return nil, err
 	}
-	if len(image.Tags) == 1 {
-		return image, c.Client.Delete(ctx, image)
-	}
 	var remainingTags []string
 
-	imageParsedTag, err := name.NewTag(imageName, name.WithDefaultRegistry(""))
+	imageParsedRef, err := name.ParseReference(imageName, name.WithDefaultRegistry(""), name.WithDefaultTag(""))
 	if err != nil {
-		return image, nil
+		return nil, err
 	}
+	tagToDelete := ""
+	repoToDelete := ""
+
+	if _, ok := imageParsedRef.(name.Digest); ok {
+		// if the image is referenced by digest, we need to delete the tag with only registry/repository
+		repoToDelete = imageParsedRef.Context().Name()
+	} else if tag, err := name.NewTag(imageName, name.StrictValidation); err == nil {
+		tagToDelete = tag.Name()
+	}
+
+	// Getting an image, auto-suffixed with :latest also returns images that don't have that tag (explicit :latest) at all, potentially yielding errors down the line
+	if tagToDelete != "" && !slices.Contains(image.Tags, tagToDelete) {
+		return image, fmt.Errorf("image tag %s not found", imageName)
+	}
+
+	if len(image.Tags) == 1 {
+		return image, c.Client.Delete(ctx, image)
+	}
+
 	for _, tag := range image.Tags {
-		if tag != imageParsedTag.Name() {
-			remainingTags = append(remainingTags, tag)
+		if tag == tagToDelete {
+			logrus.Debugf("Delete: %s", tag)
+			continue
```

Comment:
Do we actually need to delete here? Same question below regarding repo.

Reply:
You mean if we have to do something to delete the tag?
ImageDelete just updates the image with an updated list of tags (dropping the ones we matched here) and if it was the last one, we will delete the image.